### PR TITLE
detail how unsupported CRUD operations should be defined

### PR DESCRIPTION
**Pull request by kdenhartog**
Signed-off-by: Kyle Den Hartog <kyle.denhartog@mattr.global>

This is suggested additional text to address issue #149

~~I've somehow screwed up the formatting on this section too, so I'd appreciate some help learning what's the easiest way to format the HTML for PRs as well.~~

***
<a href="https://pr-preview.s3.amazonaws.com/kdenhartog/did-core/pull/150.html" title="Last updated on Feb 12, 2020, 9:54 PM UTC (1526ba9)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/did-core/150/e54cf87...kdenhartog:1526ba9.html" title="Last updated on Feb 12, 2020, 9:54 PM UTC (1526ba9)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -2211,13 +2211,7 @@ <h2>
...
 particular <a>DID registry</a>, a <a>DID method</a> specification MUST specify
 how each of the following <abbr title="Create, Read, Update, Delete">CRUD</abbr>
-operations is performed by a client. Specify each operation to the level of
-detail necessary to build and test interoperable client implementations with the
-<a>DID registry</a>. These operations can effectively be used to perform all the
-operations required of a CKMS (cryptographic key management system). For
-example, key registration, key replacement, key rotation, key recovery, and key
-expiration.
-
+operations is performed by a client. In the event that an operation is unsupported by a <a>DID method</a>, the <a>DID method</a> specification MUST specify that it's not supported. Specify each operation to the level of detail necessary to build and test interoperable client implementations with the <a>DID registry</a>. These operations can effectively be used to perform all the operations required of a CKMS (cryptographic key management system). For example, key registration, key replacement, key rotation, key recovery, and key expiration.
```
Comment (gjgd) [inline index.html]:
```suggestion
operations is performed by a client. In the event that an operation is unsupported by a <a>DID method</a>, the <a>DID method</a> specification MUST specify that it's not supported. If an operation is supported, it MUST be specified to the level of detail necessary to build and test interoperable client implementations with the <a>DID registry</a>. These operations can effectively be used to perform all the operations required of a CKMS (cryptographic key management system). For example, key registration, key replacement, key rotation, key recovery, and key expiration.
```

**Comment by kdenhartog** [comment]
I'm curious how we should account for normative statements with this. Since compliance requires "manual checks" (a person checks it), does that effect which normative statements we should be using?

**Comment by kdenhartog** [comment]
> I'm fine with merging as-is, but I also wonder if perhaps only Update and Deactivate should be optional. Otherwise, what would be the point of a DID method that can't create or resolve DIDs?
> 
> Note that currently the sections about [Update](https://w3c.github.io/did-core/#update) and [Deactivate](https://w3c.github.io/did-core/#deactivate) include the following language, which the sections about the other operations do not have:
> 
> * _...or state that updates are not possible._
> * _...or state that deactivation is not possible._
> 
> In the CCG there was an earlier PR [w3c-ccg/did-spec#243](https://github.com/w3c-ccg/did-spec/pull/243) that changed this language; also see [w3c-ccg/did-spec#55](https://github.com/w3c-ccg/did-spec/pull/55) for additional context.

Based on what I read @peacekeeper this doesn't appear to be changing anything, but rather making text consistent across different sections. I feel like I might be missing a nuance that you were trying to indicate though. From what I read in the other comments, I'm generally in agreement with what led to the merged changes. Essentially, I just want to enforce that fact the method specification editors `MUST` state they're not supporting update or delete if they aren't (like did:key did) in order to make it easier for implementors of specific methods.

File: index.html [hunk 2]
```diff
@@ -2211,13 +2211,15 @@ <h2>
...
 particular <a>DID registry</a>, a <a>DID method</a> specification MUST specify
 how each of the following <abbr title="Create, Read, Update, Delete">CRUD</abbr>
-operations is performed by a client. Specify each operation to the level of
-detail necessary to build and test interoperable client implementations with the
-<a>DID registry</a>. These operations can effectively be used to perform all the
-operations required of a CKMS (cryptographic key management system). For
-example, key registration, key replacement, key rotation, key recovery, and key
+operations is performed by a client. In the event that an operation is 
+unsupported by a <a>DID method</a>, the <a>DID method</a> specification 
+MUST specify that it's not supported. If an operation is supported, it
```
Comment (msporny) [inline index.html]:
```suggestion
MUST specify that it is not supported. If an operation is supported, it 
```

File: index.html [hunk 3]
```diff
@@ -2211,13 +2211,15 @@ <h2>
...
-operations is performed by a client. Specify each operation to the level of
-detail necessary to build and test interoperable client implementations with the
-<a>DID registry</a>. These operations can effectively be used to perform all the
-operations required of a CKMS (cryptographic key management system). For
-example, key registration, key replacement, key rotation, key recovery, and key
+operations is performed by a client. In the event that an operation is 
+unsupported by a <a>DID method</a>, the <a>DID method</a> specification 
+MUST specify that it's not supported. If an operation is supported, it 
+MUST be specified to the level of detail necessary to build and test 
+interoperable client implementations with the <a>DID registry</a>. These
```
Comment (msporny) [inline index.html]:
```suggestion
interoperable client implementations. These 
```

File: index.html (hunk 2 above)
Comment (msporny) [reply index.html]: Avoid use of contractions in spec language as contractions tend to be more difficult to read for non-native speakers and makes normative statements feel like informal dialog.

File: index.html (hunk 3 above)
Comment (msporny) [reply index.html]: Remove unnecessary end of sentence.

**Comment by msporny** [comment]
> I'm curious how we should account for normative statements with this. Since compliance requires "manual checks" (a person checks it), does that effect which normative statements we should be using?

This is a big problem with the current spec. We have effectively untestable use of normative language in the spec. We'll need to rewrite a lot of the normative language around DID Method specifications to make it less subjective. For example: 

> MUST be specified to the level of detail necessary to build and test example, key registration, key replacement, key rotation, key recovery, and key	interoperable client implementations 

Not to pick on language in this PR (because the DID Method section has lots more of it), but what exactly does "specified to the level of detail necessary" really mean... some developers can figure it out, others want it spelled out precisely.

My guess is that we're going to have to go through the entire DID Method requirements section and replace all normative language with non-normative "is expected to" "is strongly urged", or have a non-subjective checklist that you need to pass to go from "experimental" to "testing" to "production". Ideally, we have a software tool to automate the testing (and if so, we can return back to using normative language). Perhaps what we should be doing is creating a DID Method spec linting tool that searches for specific section headers, tables, content, etc.

**Comment by peacekeeper** [comment]
@kdenhartog What I meant was that your PR makes it sound like all four operations are optional for DID methods, whereas the earlier PR https://github.com/w3c-ccg/did-spec/pull/243 explicitly mentioned Update and Deactivate. But I think the individual sub-sections for the four operations clarify this further, so I'm fine with this PR the way it is now.

**Comment by msporny** [comment]
Ping @kdenhartog -- requested changes a while ago, @peacekeeper asked you some questions -- please respond to change requests by merging or rejecting, and then answer @peacekeeper, so we can finish processing this PR. 

Editors are fine to pull it in, as long as you make the changes.

**Comment by msporny** [comment]
> Editors are fine to pull it in, as long as you make the changes and then fix merge conflicts.

@kdenhartog -- merge conflicts remain, please fix.

**Comment by kdenhartog** [comment]
@msporny this should be ready now. Sorry about the delay in getting back to this.

**Comment by msporny** [comment]
Editorial, multiple reviews, comments addressed, merging.

**Pull request merged**